cowbird/handlers/impl/geoserver.py

```diff
@@ -141,38 +258,110 @@ def get_instance():
         """
         return HandlerFactory().get_handler("Geoserver")
 
+    @staticmethod
+    def publish_shapefile_task_chain(workspace_name, shapefile_name):
+        # type: (str, str) -> None
+        """
+        Applies the chain of tasks required to publish a new file to Geoserver.
+        """
+        res = chain(validate_shapefile.si(workspace_name, shapefile_name),
+                    publish_shapefile.si(workspace_name, shapefile_name))
+        res.delay()
+
     def on_created(self, filename):
         """
-        Call when a new file is found.
+        Called when a new file is found.
 
         :param filename: Relative filename of a new file
         """
-        if filename.endswith(".shp"):
+        # TODO: ajouter un case pour un workspace (folder)?
+        #  On a normalement un workspace par user et ils sont déjà créés et ajoutés à Geoserver lors de la création d'un
+        #  user. Est-ce qu'on supporte un cas où un folder serait créé manuellement? Est-ce que c'est automatiquement un
+        #  workspace à ajouter à Geoserver et à Magpie?
+        if filename.endswith(tuple(SHAPEFILE_ALL_EXTENSIONS)):
             workspace_name, shapefile_name = self._get_shapefile_info(filename)
+
             LOGGER.info("Starting Geoserver publishing process for [%s]", filename)
-            res = chain(validate_shapefile.si(workspace_name, shapefile_name),
-                        publish_shapefile.si(workspace_name, shapefile_name))
-            res.delay()
+            Geoserver.publish_shapefile_task_chain(workspace_name, shapefile_name)
+
+            self._update_magpie_layer_permissions(workspace_name, shapefile_name)
+
+    @staticmethod
+    def remove_shapefile_task(workspace_name, shapefile_name):
+        # type: (str, str) -> None
+        """
+        Applies the celery task required to remove a shapefile from Geoserver.
+        """
+        remove_shapefile.delay(workspace_name, shapefile_name)
 
     def on_deleted(self, filename):
         """
-        Call when a file is deleted.
+        Called when a file is deleted.
 
         :param filename: Relative filename of the removed file
         """
-        if filename.endswith(".shp"):
+        # TODO: voir pour le case workspace (folder)
```

Comment:
I am investigating the case of a folder deletion, and have some questions.

One thing is that with the current code, the Geoserver workspace is only removed when the user is deleted.

Also, the actual `shapefile_datastore` is the folder that is monitored. If the folder is deleted, the event will be detected here to apply any related adjustments, but recreating a folder of the same name after will not be detected by the monitoring anymore. So, we can't process any other changes on a `shapefile_datastore` folder, unless we recreate a new user.

So, if the folder is deleted, I think we should remove the geoserver workspace and unregister the monitoring to be sure there is nothing left as invalid. We would also remove corresponding Magpie resources, as discussed.

It just feels like a weird case that we wouldn't necessarily want to happen. Or maybe we should find a way to detect the creation of the `shapefile_datastore` folder if we need to reset the workspace?

Not too sure about this, and would appreciate your feedback! I hope this was clear enough, but we can talk about it  Monday if needed.

Reply:
Removing the Magpie resources will wipe out all permissions for all users/groups on the corresponding resources. If this deletion is limited only to items that reside specifically under this user's workspace, then this is fine, but for groups, it is not the same. We need to be careful not to wipe too hastly the resource hierarchy in case custom sharing permissions were applied by an admin. That would cause a lot of frustration for them.

When you say "recreating a folder of the same name after", do you mean if it is recreated manually? 
If that is the case, this is probably a case we could ignore (or just add an implementation note), since this should not be done this way (let the server manage those directories). 

If instead you meant if it is recreated also when a user is recreated, can we update the monitoring utility at the same time during that initial user-create event?

Reply:
> Removing the Magpie resources will wipe out all permissions for all users/groups on the corresponding resources. If this deletion is limited only to items that reside specifically under this user's workspace, then this is fine, but for groups, it is not the same. We need to be careful not to wipe too hastly the resource hierarchy in case custom sharing permissions were applied by an admin. That would cause a lot of frustration for them.

The case here is indeed for a folder found under a user's workspace, so that's why I assume it was correct to remove the resource, also considering we have not designed a corresponding `group` concept for the file system and that datastores are not shared (for now at least).

> When you say "recreating a folder of the same name after", do you mean if it is recreated manually?
If that is the case, this is probably a case we could ignore (or just add an implementation note), since this should not be done this way (let the server manage those directories).
If instead you meant if it is recreated also when a user is recreated, can we update the monitoring utility at the same time during that initial user-create event?

Well, I see 2 ways that the folder is created now, either manually or from the user creation. If the user is recreated, I think it should already work since it recreates the folder and registers a monitor for the folder. And for the manual creation case, as we said, it will fail, but I agree that it is a weird case that we shouldn't do.

So I was wondering if the Magpie resource should even be removed in the case of a manual folder deletion (weird case too?). Should the resource only deleted on the user deletion? Should be only display a warning/error if the folder is deleted manually if we don't want to support that case?

Reply:
If we triggered user deletion from the event when directory is manually deleted, that could potentially create a cascading effect of user, resource and permissions removal. Since user workspace direcotories should not be modified manually, I think it is safer to ignore this case and log the warning, just so we avoid too much destructive opereations on a manipulation error from an admin. If the intent was to actually remove the user, the admin would eventually do it through Magpie anyway at some point. When the user is explicitly deleted from Magpie, then it is safe to remove everything in their workspace.

Reply:
Alright, I agree that it seems more safe to only display a warning.

So, I will then only remove the workspace permissions when a `user_deleted` event is triggered, which already removes the geoserver workspace.